internal/deployer/deployer.go

```diff
@@ -1,96 +1,139 @@
+		logger:         logger,
+		repoURL:        repoURL,
+	}
 
-	tfPluginClient, err := gridDeployer.NewTFPluginClient(mnemonics, "sr25519", network, "", "", true, false)
+	projectName, err := deployer.getProjectName()
 	if err != nil {
-		return deployer{}
+		return Deployer{}, err
 	}
-	return deployer{tfPluginClient: &tfPluginClient}
+	deployer.projectName = projectName
+
+	return deployer, nil
 }
 
-func (b *deployer) Deploy(ctx context.Context, repoURL string) (string, error) {
-	network := workloads.ZNet{
-		Name:        "gridifyNetwork",
-		Description: "network for testing",
-		Nodes:       []uint32{3},
-		IPRange: gridtypes.NewIPNet(net.IPNet{
-			IP:   net.IPv4(10, 20, 0, 0),
-			Mask: net.CIDRMask(16, 32),
-		}),
-		AddWGAccess: false,
-	}
+// Deploy deploys a project and map each port to a domain
+func (d *Deployer) Deploy(ctx context.Context, ports []uint) (map[uint]string, error) {
 
-	vm := workloads.VM{
-		Name:       "gridifyVM",
-		Flist:      "https://hub.grid.tf/aelawady.3bot/abdulrahmanelawady-gridify-test-latest.flist",
-		CPU:        2,
-		PublicIP:   true,
-		Planetary:  true,
-		Memory:     1024,
-		RootfsSize: 20 * 1024,
-		Entrypoint: "/sbin/zinit init",
-		EnvVars: map[string]string{
-			"REPO_URL": repoURL,
-		},
-		NetworkName: network.Name,
-	}
+	d.logger.Debug().Msg("getting nodes with free resources")
 
-	err := b.tfPluginClient.NetworkDeployer.Deploy(ctx, &network)
+	node, err := findNode(d.tfPluginClient.GridProxyClient)
 	if err != nil {
-		return "", err
+		return map[uint]string{}, errors.Wrapf(
+			err,
+			"failed to get a node with enough resources on network %s",
+			d.tfPluginClient.Network,
+		)
 	}
 
-	dl := workloads.NewDeployment("gridifyVM", 3, "", nil, network.Name, nil, nil, []workloads.VM{vm}, nil)
-	err = b.tfPluginClient.DeploymentDeployer.Deploy(ctx, &dl)
+	d.logger.Info().Msg("deploying network")
+	network := constructNetwork(d.projectName, node)
+	err = d.tfPluginClient.NetworkDeployer.Deploy(ctx, &network)
+	if err != nil {
+		return map[uint]string{}, errors.Wrapf(err, "could not deploy network %s on node %d", network.Name, node)
+	}
 
+	d.logger.Info().Msg("deploying vm")
+	dl := constructDeployment(network.Name, d.projectName, d.repoURL, node)
+	err = d.tfPluginClient.DeploymentDeployer.Deploy(ctx, &dl)
 	if err != nil {
-		return "", err
+		return map[uint]string{}, errors.Wrapf(err, "could not deploy vm %s on node %d", dl.Name, node)
 	}
 
-	resVM, err := b.tfPluginClient.StateLoader.LoadVMFromGrid(3, "gridifyVM")
+	resVM, err := d.tfPluginClient.State.LoadVMFromGrid(node, dl.Name, dl.Name)
 	if err != nil {
-		return "", err
+		return map[uint]string{}, errors.Wrapf(err, "could not load vm %s on node %d", dl.Name, node)
 	}
 
-	publicIP := strings.Split(resVM.ComputedIP, "/")[0]
-	if !testConnection(publicIP, "22") {
-		return "", fmt.Errorf("public ip %s is not reachable", publicIP)
+	portlessBackend := constructPortlessBackend(resVM.ComputedIP)
+
+	FQDNs := make(map[uint]string, 0)
```

Comment:
what is the point of the length hint here, if anything you can pass the number of the services "ports" needed ahead of time, and it's not guaranteed to happen, it's a minor optimization (if and only if you know the number of keys of that map) that shouldn't be handled at this stage